server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -426,17 +429,66 @@ private static void addSortedBundle(Bundle bundle, Map<String, Bundle> bundles,
         Map<String, Plugin> loaded = new HashMap<>();
         Map<String, Set<URL>> transitiveUrls = new HashMap<>();
         List<Bundle> sortedBundles = sortBundles(bundles);
-
         for (Bundle bundle : sortedBundles) {
             checkBundleJarHell(JarHell.parseClassPath(), bundle, transitiveUrls);
 
             final Plugin plugin = loadBundle(bundle, loaded);
             plugins.add(new Tuple<>(bundle.plugin, plugin));
         }
 
+        loadExtensions(plugins);
         return Collections.unmodifiableList(plugins);
     }
 
+    // package-private for test visibility
+    static void loadExtensions(List<Tuple<PluginInfo, Plugin>> plugins) {
+        Map<String, List<Plugin>> extendingPlugins = plugins.stream()
+            .flatMap(t -> t.v1().getExtendedPlugins().stream().map(extendedPlugin -> Tuple.tuple(extendedPlugin, t.v2())))
+            .collect(Collectors.groupingBy(Tuple::v1, Collectors.mapping(Tuple::v2, Collectors.toList())));
+        plugins.stream().filter(t -> t.v2() instanceof ExtensiblePlugin).forEach(t -> loadExtensionsForPlugin((ExtensiblePlugin) t.v2(),
+            extendingPlugins.getOrDefault(t.v1().getName(), List.of())));
+    }
+
+    private static void loadExtensionsForPlugin(ExtensiblePlugin extensiblePlugin, List<Plugin> extendingPlugins) {
+        ExtensiblePlugin.ExtensionLoader extensionLoader = new ExtensiblePlugin.ExtensionLoader() {
+            @Override
+            public <T> Stream<T> loadExtensions(Class<T> extensionPointType) {
+                return extendingPlugins.stream().flatMap(plugin -> createExtensions(extensionPointType, plugin));
+            }
+        };
+
+        extensiblePlugin.loadExtensions(extensionLoader);
+    }
+
+    private static <T> Stream<? extends T> createExtensions(Class<T> extensionPointType, Plugin plugin) {
+        SPIClassIterator<T> classIterator = SPIClassIterator.get(extensionPointType, plugin.getClass().getClassLoader());
+        List<T> extensions = new ArrayList<>();
+        while (classIterator.hasNext()) {
+            Class<? extends T> extensionClass = classIterator.next();
+            extensions.add(createExtension(extensionClass, plugin, extensionPointType));
+        }
+        return extensions.stream();
+    }
+
+    private static <T> T createExtension(Class<? extends T> extensionClass, Plugin plugin, Class<T> extensionPointType) {
+        try {
+            try {
+                return extensionClass.getConstructor(plugin.getClass()).newInstance(plugin);
+            } catch (NoSuchMethodException e) {
```

Comment:
I think we should lookup constructors just like in `loadPlugin`, rather than using exceptions as control flow here. The suppressed exception in the case we use the zero arg constructor would be especially confusing.